File: lynq-site/src/core/permission-engine.ts
import { AppRole, appRoleFromString, FolderFeature, BudgetAuthorityPosts } from './constants';
import { UserModel, FolderMemberModel, FolderPermissionModel } from '../models/types';

export class PermissionEngine {
  readonly user: UserModel;
  readonly userFolderMemberships: FolderMemberModel[];
  readonly folderPermissions: Record<number, FolderPermissionModel[]>;
  readonly globalPermissions: FolderPermissionModel[];

  constructor(
    user: UserModel,
    userFolderMemberships: FolderMemberModel[] = [],
    folderPermissions: Record<number, FolderPermissionModel[]> = {},
    globalPermissions: FolderPermissionModel[] = []
  ) {
    this.user = user;
    this.userFolderMemberships = userFolderMemberships;
    this.folderPermissions = folderPermissions;
    this.globalPermissions = globalPermissions;
  }

  get role(): AppRole {
    // BUGFIX: previously auto-promoted any user with 1+ folder memberships
    // to forumExeccom regardless of actual DB role. Removed — role must
    // come from profiles.role only. Matches Flutter engine exactly.
    return appRoleFromString(this.user.role);
  }

  /// Check user suspension status
  get isSuspended(): boolean {
    if (this.user.status === 'suspended') return true;
    const until = this.user.suspended_until;
    if (until && new Date(until) > new Date()) return true;
    return false;
  }

  // Tier getters
  get isTier1(): boolean { return !this.isSuspended && (this.role === AppRole.chairman || this.role === AppRole.viceChairman); }
  get isTier2(): boolean { return !this.isSuspended && this.role === AppRole.coreExeccom; }
  get isTier3(): boolean { return !this.isSuspended && this.role === AppRole.forumExeccom; }
  get isTier4(): boolean { return !this.isSuspended && this.role === AppRole.panel; }
  get isTier5(): boolean { return !this.isSuspended && this.role === AppRole.restricted; }
  get isTier6(): boolean { return !this.isSuspended && this.role === AppRole.member; }

  get isAtLeastTier1(): boolean { return !this.isSuspended && this.role >= AppRole.viceChairman; }
  get isAtLeastTier2(): boolean { return !this.isSuspended && this.role >= AppRole.coreExeccom; }
  get isAtLeastTier3(): boolean { return !this.isSuspended && this.role >= AppRole.forumExeccom; }
  get isAtLeastTier4(): boolean { return !this.isSuspended && this.role >= AppRole.panel; }

  get isEffectivelyTier1(): boolean {
    return !this.isSuspended && (this.isTier1 || (this.isTier2 && !!this.user.is_sudo));
  }

  // Actions
  get canAddMembers(): boolean { return !this.isSuspended && this._isCommitteeLead; }
  get canRemoveMembers(): boolean { return !this.isSuspended && this.isAtLeastTier1; }
  get canEditMembers(): boolean { return !this.isSuspended && this.isTier1; }
  get canAssignRoles(): boolean { return !this.isSuspended && this.isAtLeastTier1; }
  get canManageFolders(): boolean { return !this.isSuspended && this.isAtLeastTier1; }
  get canManageGlobalPermissions(): boolean { return !this.isSuspended && this.role === AppRole.chairman; }
  get canManageFolderPermissions(): boolean { return !this.isSuspended && this.isAtLeastTier1; }
  get canManagePermissions(): boolean { return !this.isSuspended && (this.canManageGlobalPermissions || this.canManageFolderPermissions); }

  get _isCommitteeLead(): boolean {
    if (this.isSuspended) return false;
    if (this.isAtLeastTier2) return true; // covers Tier1 + Tier2
    if (this.isTier3) {
      const p = (this.user.post || '').toLowerCase();
      return p === 'chair' || p === 'chairman' || p === 'secretary';
    }
    return false;
  }

  get _isTop4(): boolean {
    if (this.isSuspended) return false;
    if (this.isTier1) return true;
    if (this.isTier2) {
      const p = (this.user.post || '').toLowerCase();
      return p === 'secretary' || p === 'treasurer';
    }
    return false;
  }

  canManageMembersInFolder(folderId: number): boolean {
    if (this.isSuspended) return false;
    if (this.isAtLeastTier1) return true;
    if (this.isAtLeastTier2 && this.isMemberOfFolder(folderId)) return true;
    
    // Wire up manageAll bypass
    if (this.canDoInFolder(folderId, FolderFeature.manageAll)) return true;

    // Exact folder-role match
    const fRole = (this.folderRoleIn(folderId) || '').toLowerCase().trim();
    return fRole === 'chair' || fRole === 'head';
  }

  isFeatureEnabledGlobally(feature: string): boolean {
    // BUGFIX: was reading folderPermissions[0], relying on a folder with
    // id=0 that never exists (folders PK is serial starting at 1) — this
    // always returned false. Now reads from a dedicated globalPermissions
    // map, populated separately from the global_feature_permissions table.
    if (this.isSuspended) return false;
    const perm = this.globalPermissions.find((p) => p.feature === feature);
    return perm ? perm.allowed : false;
  }

  // Feature limits
  get canCreateEvents(): boolean { return !this.isSuspended && this.isAtLeastTier3; }
  get canReadReports(): boolean { return !this.isSuspended && (this.isAtLeastTier2 || this.isFeatureEnabledGlobally(FolderFeature.viewReports)); }
  get canUploadReports(): boolean { return !this.isSuspended && this.isAtLeastTier4; }
  get canViewTotalBudget(): boolean { return !this.isSuspended && (this.isAtLeastTier2 || this.isFeatureEnabledGlobally(FolderFeature.viewTotalBudget)); }
  get canAccessScopedBudget(): boolean { return !this.isSuspended && this.isAtLeastTier3; }
  get canViewMembers(): boolean { return !this.isSuspended && this.role !== AppRole.restricted; }
  get isPanel(): boolean { return !this.isSuspended && this.isTier4; }
  get canRequestBudget(): boolean { return !this.isSuspended && this.isAtLeastTier3; }
  get canManageBudget(): boolean { return !this.isSuspended && this._isTop4; }
  get canViewInternalAnnouncements(): boolean { return !this.isSuspended && this.isAtLeastTier3; }
  get canManageAnnouncements(): boolean { return !this.isSuspended && this.isAtLeastTier2; }
  get canOverride(): boolean { return !this.isSuspended && this.role === AppRole.chairman; }
  get canAccessChat(): boolean { return !this.isSuspended && this.isAtLeastTier4; }

  get canApproveBudget(): boolean {
    if (this.isSuspended) return false;
    if (this.isEffectivelyTier1) return true;
    if (this.isTier2) return BudgetAuthorityPosts.hasBudgetAuthority(this.user.post);
    return false;
  }

  // Folder Scopes
  isMemberOfFolder(folderId: number): boolean {
    return !this.isSuspended && this.userFolderMemberships.some((m) => m.folder_id === folderId);
  }

  folderRoleIn(folderId: number): string | undefined {
    if (this.isSuspended) return undefined;
    const membership = this.userFolderMemberships.find((m) => m.folder_id === folderId);
    return membership?.folder_role;
  }

  canDoInFolder(folderId: number, feature: string): boolean {
    if (this.isSuspended) return false;
    if (this.isEffectivelyTier1) return true;

    // BUGFIX: membership check now applies to ALL tiers uniformly, including
    // Tier2. Previously Tier2 skipped this entirely and could act in folders
    // they were never added to.
    if (!this.isMemberOfFolder(folderId)) return false;

    const perms = this.folderPermissions[folderId] || [];

    // If checking regular feature, see if manageAll is explicitly allowed
    if (feature !== FolderFeature.manageAll) {
      const manageAllPerm = perms.find((p) => p.feature === FolderFeature.manageAll);
      if (manageAllPerm?.allowed) {
        return true;
      }
    }

    const perm = perms.find((p) => p.feature === feature);

    // BUGFIX: default-deny when no explicit permission row exists (was
    // default-allow for Tier2/Tier3 — meant any un-configured feature was
    // silently open). Now every feature must be explicitly granted.
    return perm ? perm.allowed : false;
  }

  canCreateEventInFolder(folderId: number): boolean {
    return !this.isSuspended && this.canCreateEvents && this.canDoInFolder(folderId, FolderFeature.createEvents);
  }

  canUploadReportInFolder(folderId: number): boolean {
    return !this.isSuspended && this.canUploadReports && this.canDoInFolder(folderId, FolderFeature.uploadReports);
  }

  canViewBudgetInFolder(folderId: number): boolean {
    return !this.isSuspended && this.canAccessScopedBudget && this.canDoInFolder(folderId, FolderFeature.viewBudget);
  }

  canRequestBudgetInFolder(folderId: number): boolean {
    return !this.isSuspended && this.canRequestBudget && this.canDoInFolder(folderId, FolderFeature.requestBudget);
  }

  get userFolderIds(): number[] {
    return this.isSuspended ? [] : this.userFolderMemberships.map((m) => m.folder_id);
  }
}

File: lynq/lib/core/permission_engine.dart
import '../models/user_model.dart';
import '../models/folder_model.dart';
import 'constants.dart';

/// Central permission resolver.
/// Checks: role level → folder membership → folder permissions → post-specific rules.
class PermissionEngine {
  final UserModel user;
  final List<FolderMemberModel> userFolderMemberships;
  final Map<int, List<FolderPermissionModel>> folderPermissions;
  final List<FolderPermissionModel> globalPermissions;

  const PermissionEngine({
    required this.user,
    this.userFolderMemberships = const [],
    this.folderPermissions = const {},
    this.globalPermissions = const [],
  });

  AppRole get role => AppRole.fromString(user.role);

  /// Check user suspension status
  bool get isSuspended {
    if (user.status == 'suspended') return true;
    final until = user.suspendedUntil;
    if (until != null && until.isAfter(DateTime.now())) return true;
    return false;
  }

  // ── Tier-based getters ──

  bool get isTier1 => !isSuspended && (role == AppRole.chairman || role == AppRole.viceChairman);
  bool get isTier2 => !isSuspended && role == AppRole.coreExeccom;
  bool get isTier3 => !isSuspended && role == AppRole.forumExeccom;
  bool get isTier4 => !isSuspended && role == AppRole.panel;
  bool get isTier5 => !isSuspended && role == AppRole.restricted;
  bool get isTier6 => !isSuspended && role == AppRole.member;

  bool get isAtLeastTier1 => !isSuspended && role >= AppRole.viceChairman;
  bool get isAtLeastTier2 => !isSuspended && role >= AppRole.coreExeccom;
  bool get isAtLeastTier3 => !isSuspended && role >= AppRole.forumExeccom;
  bool get isAtLeastTier4 => !isSuspended && role >= AppRole.panel;

  /// Effective Tier 1 (includes Tier 2 with Chairman's Sudo grant)
  bool get isEffectivelyTier1 => !isSuspended && (isTier1 || (isTier2 && user.isSudo));

  // ── Action permissions ──

  bool get canAddMembers => !isSuspended && _isCommitteeLead;
  bool get canRemoveMembers => !isSuspended && isAtLeastTier1;
  bool get canEditMembers => !isSuspended && isTier1;
  bool get canAssignRoles => !isSuspended && isAtLeastTier1;
  bool get canManageFolders => !isSuspended && isAtLeastTier1;
  bool get canManageGlobalPermissions => !isSuspended && role == AppRole.chairman;
  bool get canManageFolderPermissions => !isSuspended && isAtLeastTier1;
  bool get canManagePermissions => !isSuspended && (canManageGlobalPermissions || canManageFolderPermissions);

  /// Committee leadership: Tier 1 & 2 are always leads.
  /// Forum Execom (Tier 3) only if they hold Chair or Secretary position.
  bool get _isCommitteeLead {
    if (isSuspended) return false;
    if (isAtLeastTier2) return true; // Covers Tier 1 + Tier 2
    if (isTier3) {
      final p = user.post?.toLowerCase() ?? '';
      return p == 'chair' || p == 'chairman' || p == 'secretary';
    }
    return false;
  }

  /// Top 4 roles with full administrative and budget authority:
  /// Chairman, Vice Chairman, Secretary, Treasurer.
  bool get _isTop4 {
    if (isSuspended) return false;
    if (isTier1) return true;
    if (isTier2) {
      final p = user.post?.toLowerCase() ?? '';
      return p == 'secretary' || p == 'treasurer';
    }
    return false;
  }

  /// Can manage members within a specific folder/forum
  bool canManageMembersInFolder(int folderId) {
    if (isSuspended) return false;
    if (isAtLeastTier1) return true;
    // Core members assigned to a folder can manage its members
    if (isAtLeastTier2 && isMemberOfFolder(folderId)) return true;
    
    // Wire up manageAll bypass
    if (canDoInFolder(folderId, FolderFeature.manageAll)) return true;

    // Forum Chairs/Heads can manage their own members (exact folder-role match)
    final fRole = folderRoleIn(folderId)?.toLowerCase() ?? '';
    return fRole == 'chair' || fRole == 'head';
  }

  /// Is a specific feature allowed globally (org-wide toggle, not folder-scoped)
  bool isFeatureEnabledGlobally(String feature) {
    // BUGFIX: was reading folderPermissions[0], relying on a folder with
    // id=0 that never exists (folders PK is serial starting at 1) — this
    // always returned false. Now reads from a dedicated globalPermissions
    // list, populated from the global_feature_permissions table.
    if (isSuspended) return false;
    final perm = globalPermissions.where((p) => p.feature == feature).firstOrNull;
    return perm?.allowed ?? false;
  }

  // ── Mandated Permissions ──

  /// 1. Event creation privileges restricted exclusively to Tier 3 and above.
  bool get canCreateEvents => !isSuspended && isAtLeastTier3;

  /// 2. Report viewing access is limited to Tier 2 and all tiers above.
  bool get canReadReports => !isSuspended && (isAtLeastTier2 || isFeatureEnabledGlobally(FolderFeature.viewReports));

  bool get canUploadReports => !isSuspended && isAtLeastTier4;

  /// 5. Full organizational budget viewing access is restricted to Tier 2 and above.
  bool get canViewTotalBudget => !isSuspended && (isAtLeastTier2 || isFeatureEnabledGlobally(FolderFeature.viewTotalBudget));

  /// 6. Tier 3 (Forum-Execom) may view their forum budget. No activation gate.
  bool get canAccessScopedBudget => !isSuspended && isAtLeastTier3;

  /// Anyone EXCEPT restricted members can view the org member list (basic details only).
  bool get canViewMembers => !isSuspended && role != AppRole.restricted;

  // ── Legacy/Other Access ──
  bool get isPanel => !isSuspended && isTier4;
  bool get canRequestBudget => !isSuspended && isAtLeastTier3;
  bool get canManageBudget => !isSuspended && _isTop4;

  bool get canViewInternalAnnouncements => !isSuspended && isAtLeastTier3;
  bool get canManageAnnouncements => !isSuspended && isAtLeastTier2;
  bool get canOverride => !isSuspended && role == AppRole.chairman;
  bool get canAccessChat => !isSuspended && isAtLeastTier4;

  /// Budget approve/reject: only effectively Tier 1
  /// or Tier 2 with specific posts (Treasurer, Sub-Treasurer)
  bool get canApproveBudget {
    if (isSuspended) return false;
    if (isEffectivelyTier1) return true;
    if (isTier2) return BudgetAuthorityPosts.hasBudgetAuthority(user.post);
    return false;
  }

  // ── Folder-scoped permissions ──

  /// Is this user a member of a specific folder?
  bool isMemberOfFolder(int folderId) =>
      !isSuspended && userFolderMemberships.any((m) => m.folderId == folderId);

  /// Get the user's role within a folder
  String? folderRoleIn(int folderId) {
    if (isSuspended) return null;
    final membership = userFolderMemberships
        .where((m) => m.folderId == folderId)
        .firstOrNull;
    return membership?.folderRole;
  }

  /// Check if a folder feature is allowed for this user
  bool canDoInFolder(int folderId, String feature) {
    if (isSuspended) return false;
    // Effective Tier 1 can do everything in any folder
    if (isEffectivelyTier1) return true;

    // BUGFIX: membership check now applies uniformly to ALL tiers, including
    // Tier2. Previously Tier2 skipped this and could act in folders they
    // were never added to.
    if (!isMemberOfFolder(folderId)) return false;

    final perms = folderPermissions[folderId] ?? [];

    // If checking a regular feature, see if they have explicit manageAll allowed
    if (feature != FolderFeature.manageAll) {
      final manageAllPerm = perms.where((p) => p.feature == FolderFeature.manageAll).firstOrNull;
      if (manageAllPerm?.allowed ?? false) {
        return true;
      }
    }

    final perm = perms.where((p) => p.feature == feature).firstOrNull;

    // BUGFIX: default-deny when no explicit permission row exists (was
    // default-allow for Tier2/Tier3 — any un-configured feature was
    // silently open). Every feature must be explicitly granted now.
    return perm?.allowed ?? false;
  }

  /// Can create events in a specific folder
  bool canCreateEventInFolder(int folderId) =>
      !isSuspended && canCreateEvents && canDoInFolder(folderId, FolderFeature.createEvents);

  /// Can upload reports in a specific folder
  bool canUploadReportInFolder(int folderId) =>
      !isSuspended && canUploadReports && canDoInFolder(folderId, FolderFeature.uploadReports);

  /// Can view budget in a specific folder
  bool canViewBudgetInFolder(int folderId) =>
      !isSuspended && canAccessScopedBudget && canDoInFolder(folderId, FolderFeature.viewBudget);

  /// Can request budget in a specific folder (wired up)
  bool canRequestBudgetInFolder(int folderId) =>
      !isSuspended && canRequestBudget && canDoInFolder(folderId, FolderFeature.requestBudget);

  /// Get list of folder IDs this user belongs to
  List<int> get userFolderIds =>
      isSuspended ? [] : userFolderMemberships.map((m) => m.folderId).toList();
}
